core/src/floats.rs

```diff
@@ -33,58 +111,69 @@ impl<T1: Datum + Float, T2: Datum + Float>
 {
     fn translate_node(
         &self,
-        _source: &Graph<TypedFact, Box<dyn TypedOp>>,
+        source: &Graph<TypedFact, Box<dyn TypedOp>>,
         node: &Node<TypedFact, Box<dyn TypedOp>>,
         target: &mut Graph<TypedFact, Box<dyn TypedOp>>,
         mapping: &HashMap<OutletId, OutletId>,
     ) -> TractResult<TVec<OutletId>> {
-        let new_op = if let Some(source) = node.op_as::<TypedSource>() {
-            Box::new(TypedSource::new(fact_float_precision_conversion::<T1, T2>(&source.fact)))
-        } else if let Some(konst) = node.op_as::<Const>() {
-            Box::new(Const(tensor_float_precision_conversion::<T1, T2>(&konst.0)))
-        } else if let Some(cast) = node.op_as::<Cast>() {
-            if cast.to == T1::datum_type() {
-                Box::new(Cast { to: T2::datum_type() })
-            } else {
-                node.op.clone()
-            }
-        } else if let Some(ew) = node.op_as::<ElementWiseOp>() {
-            if ew.1 == Some(T1::datum_type()) {
-                Box::new(ElementWiseOp(ew.0.clone(), Some(T2::datum_type())))
-            } else {
-                node.op.clone()
-            }
-        } else if let Some(bin) = node.op_as::<TypedBinOp>() {
-            if bin.1 == Some(T1::datum_type()) {
-                Box::new(TypedBinOp(bin.0.clone(), Some(T2::datum_type())))
-            } else {
-                node.op.clone()
-            }
-        } else if let Some(op) = node.op_as::<Scan>() {
-            let body = FloatPrecisionTranslator::<T1, T2>::default().translate_model(&op.body)?;
-            Box::new(Scan { body, ..op.clone() })
-        } else if let Some(op) = node.op_as::<EinSum>() {
-            Box::new(EinSum {
-                operating_dt: dt_float_precision_conversion::<T1, T2>(op.operating_dt),
-                ..op.clone()
-            })
-        } else if let Some(op) = node.op_as::<Pad>() {
-            if let PadMode::Constant(t) = &op.mode {
-                Box::new(Pad {
-                    mode: PadMode::Constant(tensor_float_precision_conversion::<T1, T2>(t)),
+        let is_source = node.op_as::<TypedSource>().is_some();
```

Comment:
Not sure why Source need a special treatment ?

Reply:
I am treating the model input differently in order to preserve that input and ouput at the model level are still f16 after a f32-to-f16 transformation. That is also the purpose of the function `cast_model_outputs_if_required ` but on the model output side. 

If a source is needed in f32 from an input of f16, cast operation will be added.